src/Compilers/Core/Portable/SourceGeneration/SyntaxContextReceiverAdaptor.cs

```diff
@@ -11,13 +11,24 @@ namespace Microsoft.CodeAnalysis
     /// </summary>
     internal sealed class SyntaxContextReceiverAdaptor : ISyntaxContextReceiver
     {
-        public SyntaxContextReceiverAdaptor(ISyntaxReceiver? receiver)
+        private SyntaxContextReceiverAdaptor(ISyntaxReceiver receiver)
         {
-            Receiver = receiver;
+            Receiver = receiver ?? throw new ArgumentNullException(nameof(receiver));
         }
 
-        public ISyntaxReceiver? Receiver { get; }
+        public ISyntaxReceiver Receiver { get; }
 
-        public void OnVisitSyntaxNode(GeneratorSyntaxContext context) => Receiver?.OnVisitSyntaxNode(context.Node);
+        public void OnVisitSyntaxNode(GeneratorSyntaxContext context) => Receiver.OnVisitSyntaxNode(context.Node);
+
+        public static SyntaxContextReceiverCreator Create(SyntaxReceiverCreator creator) => () =>
+        {
+            var rx = creator();
+            if (rx is object)
+            {
+                return new SyntaxContextReceiverAdaptor(rx);
+            }
+            // in the case that the creator function returns null, we'll also return a null adaptor
+            return null!;
```

Comment:
It seems surprising to return `null!`. Consider checking the `creator` in the caller of this method and only calling when not `null`.